### PR TITLE
[BI-110] Implemented Entity Deletion Tracking Support

### PR DESCRIPTION
**Pull request by full-of-foo**
TL;DR: currently only entity 'creates' and 'updates' are supported in both the full and incremental extract tasks/processes. Here entity deletion is supported based on `DELETED` change tracking events.

Ticket: https://fieldaware.atlassian.net/browse/BI-110
 
Tasks:
 - Enhanced Dockerfile (stricter and more verbose commands through 'set' chains)
 - Kill node process once all task promises resolve
 - Do not use 'minimize' in all Mongoose schemas to ensure that no null/empty data attributes are trimed when being inserted into documents
 - Add a config module for listing configs based or current node env (dev and test)
 - Added deletion tracking support in onSuccess of TrackingRequestFactory
 - Use less verbose arrow function syntax in places
 - Added nock dep for mocking requests
 - Added asset, user and tracking JSON fixtures
 - Added requestFactory fixture
 - Added naive ModelSpec
 - Added basic EntityRequestFactorySpec and TrackingRequestFactorySpec

Note: 
 - This repo originates from a spike (https://fieldaware.atlassian.net/wiki/display/BI/%5BSpike%5D%5BDL%5D+Lightweight+Mongo+DW+replacement+for+ADS+with+naive+extract+implementation) and only one of its PRs has been reviewed before
 - For those interested, a peer review of the entire repo is due to happen soon where wide-reaching implementation enhancements will be discussed and DLd (https://fieldaware.atlassian.net/browse/BI-107) 
 - Much of the additions here are around supporting tests for tracking requests and models (fixtures + new specs)

Reviewers:
 - [ ] @ldoran 
 - [ ] @marzetas

### CODE REVIEW

**Comment by marzetas** [comment]
Just for the record, considering the improvements done by github on PRs, next time try to put the results of each task in an atomic commit. It makes it easier to review, and also to revert if anything is wrong.

File: fa-dw/app/lib/request/TrackingRequestFactory.js [hunk 1]
```diff
@@ -19,27 +20,40 @@ let _requestNextTrackingPage = (url, qs, Model, bt) => {
...
+
+class TrackingRequestFactory extends RequestFactory {
   onSuccess(res, body) {
-    let pageEntityUrls = body.items.map((item) => { return item.url; });
     let entityReqQs = { pageSize: constants.DEFAULT_PAGE_SIZE };
-    // let trackingPagePromises = [];
+    let reqEntityItems = body.items.filter((item) => item.status === 'UPDATED' || item.status === 'CREATED');
+    let deleteEntityItems = body.items.filter((item) => item.status === 'DELETED' );
+    let reqEntityUrls = reqEntityItems.map((item) => item.url );
+    let deleteEntityIds = deleteEntityItems.map((item) => url.parse(item.url).pathname.split('/')[2] );
```
Comment (marzetas) [inline fa-dw/app/lib/request/TrackingRequestFactory.js]: Can't this be done in `map/reduce` fashion, instead of filter/map?

File: fa-dw/app/lib/tasks.js [hunk 2]
```diff
@@ -15,6 +15,11 @@ let TrackingRequestFactory = require('./request/TrackingRequestFactory');
 let TokenRequestFactory = require('./request/TokenRequestFactory');
 
 
+let _runModelTracking = (initialUrl, bt, Model) => {
+  let qs = { entityClass: Model.modelName };
+  return new TrackingRequestFactory(initialUrl, qs, Model, bt).createRequest();
+};
```
Comment (marzetas) [inline fa-dw/app/lib/tasks.js]: Might be worth use slightly more descriptive name, I can't help asking myself what `bt` (or even `qs`) stands for...

**Comment by marzetas** [comment]
I can wait for that peer-review, I still feel like I'm reviewing things a bit out of context. 

Btw, are you following any es6 coding standards in particular? I remember you asked about it. Just curious because of the complete disappearance of `var`s, considering that in most contexts are equivalent to `let`.

File: fa-dw/app/lib/request/TrackingRequestFactory.js (hunk 1 above)
Comment (full-of-foo) [inline fa-dw/app/lib/request/TrackingRequestFactory.js]: I originally tried this and didn't have much success. I could do two reduces? Though I am not sure which approach is more expressive. Your call :heart: 

```javascript
// fyi, not tested. Should work though.
reqEntityUrls = body.items.reduce((entityUrls, item, index, items) => {
  if(item.status === 'UPDATED' || item.status === 'CREATED') entityUrls.push(item.url);
  return entityUrls;
}, []);
deleteEntityIds = body.items.reduce((entityIds, item, index, items) => {
  if(item.status === 'DELETED') entityIds.push(url.parse(item.url).pathname.split('/')[2]);
  return entityIds;
}, []);
```

**Comment by full-of-foo** [comment]
@marzetas Indeed, me too. I can only imagine, its effectively reviewing 'in the dark'. 

I haven't been formally any guide as of yet tbh. I followed up on the status of our Es6 style-guide, as it stands we haven't ours written out but we are following the Air BnB one. 

Stemming from the peer review, I shall add a task for applying that style guide across the code base :smile:

File: fa-dw/app/lib/request/TrackingRequestFactory.js [hunk 3]
```diff
@@ -37,10 +37,14 @@ let _deleteEntities = (Model, ids) => {
...
-    let reqEntityUrls = reqEntityItems.map((item) => item.url );
-    let deleteEntityIds = deleteEntityItems.map((item) => url.parse(item.url).pathname.split('/')[2] );
+    let reqEntityUrls = body.items.reduce((entityUrls, item, index, items) => {
+    if(item.status === 'UPDATED' || item.status === 'CREATED') entityUrls.push(item.url);
+      return entityUrls;
+    }, []);
+    let deleteEntityIds = body.items.reduce((entityIds, item, index, items) => {
+      if(item.status === 'DELETED') entityIds.push(url.parse(item.url).pathname.split('/')[2]);
+      return entityIds;
+    }, []);
```
Comment (marzetas) [inline fa-dw/app/lib/request/TrackingRequestFactory.js]: I would indent the body of the arrow function body.

**Comment by marzetas** [comment]
Apart from the indentation issue mentioned above, :sheep: it!
